cirq-core/cirq/transformers/measurement_transformers.py

```diff
@@ -0,0 +1,175 @@
+    """
+
+    def __init__(self, key: Union[str, 'cirq.MeasurementKey'], qid: 'cirq.Qid'):
+        """Initializes the qubit.
+
+        Args:
+            key: The key of the measurement gate being deferred.
+            qid: One qubit that is being measured. Each deferred measurement
+                should create one new _MeasurementQid per qubit being measured
+                by that gate.
+        """
+        self._key = value.MeasurementKey.parse_serialized(key) if isinstance(key, str) else key
+        self._qid = qid
+
+    @property
+    def dimension(self) -> int:
+        return self._qid.dimension
+
+    def _comparison_key(self) -> Any:
+        return (str(self._key), self._qid._comparison_key())
+
+    def __str__(self) -> str:
+        return f"M('{self._key}', q={self._qid})"
+
+    def __repr__(self) -> str:
+        return f'_MeasurementQid({self._key!r}, {self._qid!r})'
+
+
+@transformer_api.transformer
+def defer_measurements(
+    circuit: 'cirq.AbstractCircuit', *, context: Optional['cirq.TransformerContext'] = None
+) -> 'cirq.Circuit':
+    """Implements the Deferred Measurement Principle.
+
+    Uses the Deferred Measurement Principle to move all measurements to the
+    end of the circuit. All non-terminal measurements are changed to
+    conditional quantum gates onto ancilla qubits, and classically controlled
+    operations are transformed to quantum controls from those ancilla qubits.
+    Finally, measurements of all ancilla qubits are appended to the end of the
+    circuit.
+
+    Optimizing deferred measurements is an area of active research, and future
+    iterations may contain optimizations that reduce the number of ancilla
+    qubits, so one should not depend on the exact shape of the output from this
+    function. Only the logical equivalence is guaranteed to remain unchanged.
+    Moment and subcircuit structure is not preserved.
+
+    Args:
+        circuit: The circuit to transform. It will not be modified.
+        context: `cirq.TransformerContext` storing common configurable options
+            for transformers.
+    Returns:
+        A circuit with equivalent logic, but all measurements at the end of the
+        circuit.
+    Raises:
+        ValueError: If sympy-based classical conditions are used, or if
+            conditions based on multi-qubit measurements exist. (The latter of
+            these is planned to be implemented soon).
+    """
+
+    context1 = context or transformer_api.TransformerContext()
+    circuit = transformer_primitives.unroll_circuit_op(circuit, deep=True, tags_to_check=None)
+    terminal_measurements = {op for _, op in find_terminal_measurements(circuit)}
+    measurement_qubits: Dict['cirq.MeasurementKey', List['_MeasurementQid']] = {}
+
+    def defer(op: 'cirq.Operation', _) -> 'cirq.OP_TREE':
+        if any(t in context1.tags_to_ignore for t in op.tags) or op in terminal_measurements:
```

Comment:
After https://github.com/quantumlib/Cirq/pull/4933, `any(t in context1.tags_to_ignore for t in op.tags)` can now be replaced by adding `tags_to_ignore=context1.tags_to_ignore` parameter to the `map_operations_and_unroll` method.